opentelemetry-sdk/src/opentelemetry/sdk/_metrics/instrument.py

```diff
@@ -30,64 +28,41 @@
 from opentelemetry._metrics.instrument import (
     ObservableUpDownCounter as APIObservableUpDownCounter,
 )
-from opentelemetry._metrics.instrument import TCallback
 from opentelemetry._metrics.instrument import UpDownCounter as APIUpDownCounter
+from opentelemetry.sdk._metrics.aggregation import (
+    ExplicitBucketHistogramAggregation,
+    LastValueAggregation,
+    SumAggregation,
+)
+from opentelemetry.sdk._metrics.measurement import Measurement
 from opentelemetry.sdk.util.instrumentation import InstrumentationInfo
 
+TCallback = Union[
+    Callable[[], Iterable[Measurement]],
+    Generator[Iterable[Measurement], None, None],
+]
 
-class _Instrument:
-    def __init__(
-        self,
-        instrumentation_info: InstrumentationInfo,
-        name: str,
-        *args,
-        unit: str = "",
-        description: str = "",
-    ):
-        self._instrumentation_info = instrumentation_info
-        super().__init__(name, *args, unit=unit, description=description)
 
-
-class Counter(_Instrument, APICounter):
-    def __init__(
-        self,
-        instrumentation_info: InstrumentationInfo,
-        name: str,
-        unit: str = "",
-        description: str = "",
-    ):
-        super().__init__(
-            instrumentation_info,
-            name,
-            unit=unit,
-            description=description,
-        )
-
-    def add(self, amount, attributes=None):
-        # FIXME check that the amount is non negative
+class _Instrument(ABC):
+    @property
+    @abstractmethod
+    def _default_aggregation(self):
         pass
 
 
-class UpDownCounter(_Instrument, APIUpDownCounter):
+class _Synchronous(_Instrument):
```

Comment:
it is not clear to me the motive behind this and `Synchronous` in metric api package?

Reply:
`_Synchronous` and `_Asynchronous` are there to avoid repeating the `__init__` code in the SDK instruments, there is not hard relationship between the `Synchronous` and `Asynchronous` classes in the API.

Reply:
The class relationship is a bit confusing to me. What is the purpose of having `_Synchronous` when we can put the logic in `Synchronous` package?

Take an SDK `UpDownCounter` for instance. It is `_Synchronous`, and an `APIUpDownCounter` which is in turn `_NonMonotonic` and `Synchronous`. Is there logic in the sdk synch that cannot be encapsulated and applied to the api synch?

Reply:
The most clear example is the SDK `_Asynchronous` class, it has code that can not be placed in the API `Asynchronous` class, because the way we handle generator and callable callbacks is specific to the SDK. The `_Asynchronous` and `_Synchronous` classes are private and implementation details, to determine if an instrument is synchronous or not, the user uses the API `Synchronous` or `Asynchronous` classes.